### PR TITLE
Progress Reporting Blob Downloads

### LINKED ISSUE #8506: Storage: Add progress handling for upload/download
BlobBaseClient.StagedDownloadAsync and BlobClient.StagedUploadAsync have commented out progress handlers.

### PR DESCRIPTION
**Pull request by jaschrep-msft**
`BlobBaseClient` methods `DownloadStreaming()`, `DownloadContent()`, and `DownloadTo()` now accept `IProgress<long>` for reporting download progress.

While `IProgress<long>` isn't necessary for `DownloadStreaming()`, as the caller can easily report progress as they read from the result stream, it is necessary for `DownloadContent()`. Since these two methods share the same options bag, the parameter has been implemented for both.

Resolves #8506

### CODE REVIEW

File: sdk/storage/Azure.Storage.Blobs/src/BlobBaseClient.cs [hunk 1]
```diff
@@ -2063,7 +2065,7 @@ private void SetNameFieldsIfNull()
             return StagedDownloadAsync(
                 destination,
                 options?.Conditions,
-                //options.ProgressHandler, // TODO: #8506
+                options.ProgressHandler,
```
Comment (seanmcc-msft) [inline sdk/storage/Azure.Storage.Blobs/src/BlobBaseClient.cs:2068]: Does this need to be `options?.ProgressHandler`?  Same for the instances below.

File: sdk/storage/Azure.Storage.Blobs/tests/ProgressHandlingTests.cs [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+    public class ProgressHandlingTests : BlobTestBase
+    {
+        public ProgressHandlingTests(bool async, BlobClientOptions.ServiceVersion serviceVersion)
+            : base(async, serviceVersion, RecordedTestMode.Record /* RecordedTestMode.Record /* to re-record */)
+        {
+        }
+
+        [RecordedTest]
+        [TestCase(Constants.KB)] // simple case
+        [TestCase(256 * Constants.KB)] // multi-report case
```
Comment (seanmcc-msft) [inline sdk/storage/Azure.Storage.Blobs/tests/ProgressHandlingTests.cs:27]: Did we record 256 KB of data here?  Might not matter but seems like a lot.

**Review by amnguye — APPROVED** [review]

File: sdk/storage/Azure.Storage.Blobs/tests/ProgressHandlingTests.cs (hunk 2 above)
Comment (jaschrep-msft) [reply sdk/storage/Azure.Storage.Blobs/tests/ProgressHandlingTests.cs:27]: Without a way to control progress reporting increments like we can with storage transfer options, this number isn't too big while guaranteeing beyond shadow of a doubt we hit incrementing reports in a test.

**Review by seanmcc-msft — APPROVED** [review]

**Comment by rozeboosje** [comment]
great news and thank you all, folks. Let us know what build we can expect this to appear in.

**Pull request merged**